app/forms/declaration/countries/Countries.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package forms.declaration.countries
+
+import forms.DeclarationPage
+import models.DeclarationType._
+import models.requests.JourneyRequest
+import play.api.data.{Form, Forms, Mapping}
+import play.api.data.Forms.{optional, text}
+import play.api.mvc.AnyContent
+import services.Countries.allCountries
+
+object Countries {
+
+  sealed trait CountryPage extends DeclarationPage {
+    val id: String
+  }
+
+  case object OriginationCountryPage extends CountryPage {
+    override val id = "originationCountry"
+  }
+
+  case object DestinationCountryPage extends CountryPage {
+    override val id = "destinationCountry"
+  }
+
+  case object FirstRoutingCountryPage extends CountryPage {
+    override val id = "firstRoutingCountry"
+  }
+
+  case object NextRoutingCountryPage extends CountryPage {
+    override val id = "routingCountry"
+  }
+
+  private def mapping(page: CountryPage, cachedCountries: Seq[Country] = Seq.empty): Mapping[String] =
+    text()
+      .verifying(s"declaration.${page.id}.empty", _.trim.nonEmpty)
+      .verifying(s"declaration.${page.id}.error", emptyOrValidCountry)
+      .verifying(s"declaration.routingCountries.duplication", !cachedCountries.flatMap(_.code).contains(_))
```

Comment:
I approve of this file very much.   Lots of good practices for optional/mandatary mapping and supporting multiple pages.   I like it.

Only comment is about that "duplication" check.  I can see what it's doing but the not combined with the flatmap jars a bit.  Only a nit though

Reply:
This class is the old one, we had it before, just because of changing the name and more logic GitHub recognise it as new file.

We're storing the Country as a model, but the frontend autocomplete contains only country code.  This is the reason that I need map to the code